### DISCUSSION TITLE
reference data label for code assembler in 32bit mode please...

### DISCUSSION THREAD

**Discussion by wongkeli** [icedland/iced#303] [Q&A] [answered]
label data db, dd ... do not works in 32bit mode

Below code works fine in x64 but not in x86

        // ecx/rcx = adresse de l'image
        // edx/rdx = longueur ( # de points )
        var data1 = asm.CreateLabel();
        var lbFin = asm.CreateLabel();

        if (x64)
          asm.or(rdx, rdx);  // la taille
        else
          asm.or(edx, edx);
        asm.jz(lbFin);

        asm.pxor(xmm0, xmm0);
        asm.movups(xmm7, __dword_ptr[data1]);  // DO NOT WORKS IN 32-bit MODE  ( please HELP ! )

        asm.AnonymousLabel(); // @boucle:
        if (x64) {
          asm.movd(xmm1, __[rcx]);
        } else {
          asm.movd(xmm1, __[ecx]);
        }
        asm.punpcklbw(xmm1, xmm0);  // 8 bits -> 16 bits
        asm.punpcklwd(xmm1, xmm0);  // 16 bits -> 32 bits
        asm.cvtdq2ps(xmm1, xmm1);   // convert 4 integers -> 4 singles
        asm.mulps(xmm1, xmm7);      // r * 0.3, g * 0.59, b * 0.11
        asm.pshufd(xmm2, xmm1, 0b00000001);  // green dans xmm2
        asm.pshufd(xmm3, xmm1, 0b00000010);  // red dans xmm3
        asm.addss(xmm1, xmm2);               // (r + g + b)
        asm.addss(xmm1, xmm3);
        asm.pshufd(xmm1, xmm1, 0b11000000);  // valeur dans RGB + l'alpha

        asm.cvtps2dq(xmm1, xmm1);  // single => 32 integer
        asm.packssdw(xmm1, xmm1);  // 32 bits => 16 bits
        asm.packuswb(xmm1, xmm1);  // 16 bits => 8 bits = ( 32 bits )
        if (x64) {
          asm.movd(__[rcx], xmm1);
          asm.add(rcx, 4);
          asm.dec(rdx);
        } else {
          asm.movd(__[ecx], xmm1);
          asm.add(ecx, 4);
          asm.dec(edx);
        }
        asm.jnz(asm.@B);  // jnz boucle

        asm.Label(ref lbFin);  //@fin:
        asm.ret();

        asm.Label(ref data1);  //@data:
        asm.dd(0.11f, 0.59f, 0.30f, 1f);

**Comment by wtfsck** [comment, marked as answer]
See workaround for 32-bit code here: https://github.com/icedland/iced/discussions/279 Even though it's for 32-bit mode, it works in 64-bit mode too, just change the 32-bit reg to a 64-bit reg.
